src/app/AttributePathExpandIterator-DataModel.cpp

```diff
@@ -0,0 +1,243 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <app/AttributePathExpandIterator-DataModel.h>
+#include <app/GlobalAttributes.h>
+
+using namespace chip::app::InteractionModel;
+
+namespace chip {
+namespace app {
+
+AttributePathExpandIteratorDataModel::AttributePathExpandIteratorDataModel(
+    InteractionModel::DataModel * dataModel, SingleLinkedListNode<AttributePathParams> * attributePath) :
+    mDataModel(dataModel),
+    mpAttributePath(attributePath), mOutputPath(kInvalidEndpointId, kInvalidClusterId, kInvalidAttributeId)
+
+{
+    mOutputPath.mExpanded = true; // this is reset in 'next' if needed
+
+    // Make the iterator ready to emit the first valid path in the list.
+    // TODO: the bool return value here is completely unchecked
+    Next();
+}
+
+bool AttributePathExpandIteratorDataModel::IsValidAttributeId(AttributeId attributeId)
+{
+    switch (attributeId)
+    {
+    case Clusters::Globals::Attributes::GeneratedCommandList::Id:
+    case Clusters::Globals::Attributes::AcceptedCommandList::Id:
+#if CHIP_CONFIG_ENABLE_EVENTLIST_ATTRIBUTE
+    case Clusters::Globals::Attributes::EventList::Id:
+#endif // CHIP_CONFIG_ENABLE_EVENTLIST_ATTRIBUTE
+    case Clusters::Globals::Attributes::AttributeList::Id:
+        return true;
+    default:
+        break;
+    }
+
+    const ConcreteAttributePath attributePath(mOutputPath.mEndpointId, mOutputPath.mClusterId, attributeId);
+    return mDataModel->GetAttributeInfo(attributePath).has_value();
+}
+
+std::optional<AttributeId> AttributePathExpandIteratorDataModel::NextAttributeId()
+{
+    if (mOutputPath.mAttributeId == kInvalidAttributeId)
+    {
+        if (mpAttributePath->mValue.HasWildcardAttributeId())
+        {
+            AttributeEntry entry = mDataModel->FirstAttribute(mOutputPath);
+            return entry.IsValid()                                         //
+                ? entry.path.mAttributeId                                  //
+                : Clusters::Globals::Attributes::GeneratedCommandList::Id; //
+        }
+
+        // We allow fixed attribute IDs if and only if they are valid:
+        //    - they may be GLOBAL attributes OR
+        //    - they are valid attributes for this cluster
+        if (IsValidAttributeId(mpAttributePath->mValue.mAttributeId))
+        {
+            return mpAttributePath->mValue.mAttributeId;
+        }
+
+        return std::nullopt;
+    }
+
+    // advance the existing attribute id if it can be advanced
+    VerifyOrReturnValue(mpAttributePath->mValue.HasWildcardAttributeId(), std::nullopt);
+
+    // NOTE: this switch matches the order in GlobalAttributesNotInMetadata so that order-dependent
+    //       checks match
```

Comment:
How about implementing this as a loop over `GlobalAttributesNotInMetadata` so we don't have multiple sources of truth?  The whole point of GlobalAttributesNotInMetadata was to have one source of truth for that set.

Reply:
Updated.